File: py_scripts/simple_sub.py
#!/usr/bin/python
import dps
import sys
import time

key_id = [
    [0xed,0x54,0x14,0xa8,0x5c,0x4d,0x4d,0x15,0xb6,0x9f,0x0e,0x99,0x8a,0xb1,0x71,0xf2],
    [0x53,0x4d,0x2a,0x4b,0x98,0x76,0x1f,0x25,0x6b,0x78,0x3c,0xc2,0xf8,0x12,0x90,0xcc],
]

# Pre-shared keys for testing only. DO NOT USE THESE KEYS IN A REAL APPLICATION!
key_data = [
    [0x77,0x58,0x22,0xfc,0x3d,0xef,0x48,0x88,0x91,0x25,0x78,0xd0,0xe2,0x74,0x5c,0x10],
    [0x39,0x12,0x3e,0x7f,0x21,0xbc,0xa3,0x26,0x4e,0x6f,0x3a,0x21,0xa4,0xf1,0xb5,0x98],
]
network_key = [
    0xcd,0xfe,0x31,0x59,0x70,0x5f,0xe4,0xc8,0xcb,0x40,0xac,0x69,0x9c,0x06,0x3a,0x1d
]

key_store = dps.create_memory_key_store()
for i in xrange(len(key_id)):
    dps.set_content_key(key_store, key_id[i], key_data[i])
dps.set_network_key(key_store, network_key)

def on_pub(sub, pub, payload):
    print "Pub %s(%d) matches:" % (dps.publication_get_uuid(pub), dps.publication_get_sequence_num(pub))
    print "  pub " + " | ".join(dps.publication_get_topics(pub))
    print "  sub " + " | ".join(dps.subscription_get_topics(sub))
    print payload
    if dps.publication_is_ack_requested(pub):
        ack_msg = "This is an ACK from %d" % (dps.get_port_number(dps.publication_get_node(pub)))
        print "Sending ack for pub UUID %s(%d)" % (dps.publication_get_uuid(pub), dps.publication_get_sequence_num(pub))
        print "    %s" % (ack_msg)
        dps.ack_publication(pub, ack_msg);

# Enable or disable (default) DPS debug output
dps.cvar.debug = False

node = dps.create_node("/", dps.memory_key_store_handle(key_store), key_id[0])
dps.start_node(node, dps.MCAST_PUB_ENABLE_RECV, 0)
sub = dps.create_subscription(node, ['a/b/c']);
dps.subscribe(sub, on_pub)

if not sys.flags.interactive:
    while True:
        time.sleep(1)

File: go/examples/simple_sub/simple_sub.go
package main

import (
	"dps"
	"dps/examples/keys"
	"flag"
	"fmt"
	"strings"
	"time"
)

var (
	debug = flag.Bool("d", false, "enable debug output if built for debug")
	encryption = flag.Int("x", 1, "disable (0) or enable symmetric encryption (1), asymmetric encryption (2), or authentication (3)")
	network = flag.String("n", "udp", "Network of listen and link addresses")
)

func main() {
	flag.Parse()
	if *debug {
		dps.SetDebug(1)
	} else {
		dps.SetDebug(0)
	}
	var nodeId []byte
	keyStore := dps.CreateMemoryKeyStore()
	dps.SetNetworkKey(keyStore, keys.NetworkKeyId, keys.NetworkKey)
	if *encryption == 0 {
		nodeId = nil
	} else if *encryption == 1 {
		for i := 0; i < len(keys.KeyId); i++ {
			dps.SetContentKey(keyStore, keys.KeyId[i], keys.KeyData[i])
		}
		nodeId = nil
	} else if *encryption == 2 {
		dps.SetTrustedCA(keyStore, keys.CA)
		dps.SetCertificate(keyStore, keys.SubscriberCert, &keys.SubscriberPrivateKey, &keys.SubscriberPassword)
		dps.SetCertificate(keyStore, keys.PublisherCert, nil, nil)
		nodeId = []byte(keys.SubscriberId)
	}

	node := dps.CreateNode("/", keyStore, nodeId)
	addr := dps.CreateAddress()
	dps.SetAddress(addr, network, nil)
	dps.StartNode(node, dps.MCAST_PUB_ENABLE_RECV, addr)
	dps.DestroyAddress(addr)
	fmt.Printf("Subscriber is listening on %v\n", dps.GetListenAddressString(node))

	sub := dps.CreateSubscription(node, []string{"a/b/c"})
	dps.Subscribe(sub, func(sub *dps.Subscription, pub *dps.Publication, payload []byte) {
		fmt.Printf("Pub %v(%v) matches:\n", dps.PublicationGetUUID(pub), dps.PublicationGetSequenceNum(pub))
		fmt.Printf("  pub %v\n", strings.Join(dps.PublicationGetTopics(pub), " | "))
		fmt.Printf("  sub %v\n", strings.Join(dps.SubscriptionGetTopics(sub), " | "))
		fmt.Printf("%v\n", string(payload))
		if dps.PublicationIsAckRequested(pub) {
			ackMsg := fmt.Sprintf("This is an ACK from %v", dps.GetListenAddressString(node))
			fmt.Printf("Sending ack for pub UUID %v(%v)\n", dps.PublicationGetUUID(pub), dps.PublicationGetSequenceNum(pub))
			fmt.Printf("    %v\n", ackMsg)
			dps.AckPublication(pub, []byte(ackMsg))
		}
	})

	for {
		time.Sleep(1 * time.Second)
	}
}
